benchmarks/go_benchmark_test.go

```diff
@@ -0,0 +1,133 @@
+package main
```

Comment:
This benchmark can be more simply written as the following:
```go
package main

type Dataset struct {
	name        string
	newMessage  func() proto.Message
	marshaled   [][]byte
	unmarshaled []proto.Message
}

var datasets []Dataset

func init() {
	for _, f := range flag.Args() {
		// Load the benchmark.
		b, err := ioutil.ReadFile(f)
		if err != nil {
			panic(err)
		}

		// Parse the benchmark.
		var dm benchmarkWrapper.BenchmarkDataset
		if err := proto.Unmarshal(b, &dm); err != nil {
			panic(err)
		}

		// Determine the concrete protobuf message type to use.
		var ds Dataset
		switch dm.MessageName {
		case "benchmarks.proto3.GoogleMessage1":
			ds.newMessage = func() proto.Message { return new(googleMessage1Proto3.GoogleMessage1) }
		case "benchmarks.proto2.GoogleMessage1":
			ds.newMessage = func() proto.Message { return new(googleMessage1Proto2.GoogleMessage1) }
		case "benchmarks.proto2.GoogleMessage2":
			ds.newMessage = func() proto.Message { return new(googleMessage2.GoogleMessage2) }
		case "benchmarks.google_message3.GoogleMessage3":
			ds.newMessage = func() proto.Message { return new(googleMessage3.GoogleMessage3) }
		case "benchmarks.google_message4.GoogleMessage4":
			ds.newMessage = func() proto.Message { return new(googleMessage4.GoogleMessage4) }
		default:
			panic("unknown message type: " + dm.MessageName)
		}

		// Unmarshal each test message.
		for _, payload := range dm.Payload {
			ds.marshaled = append(ds.marshaled, payload)
			m := ds.newMessage()
			if err := proto.Unmarshal(payload, m); err != nil {
				panic(err)	
			}
			ds.unmarshaled = append(ds.unmarshaled, m)
		}

		datasets = append(datasets, ds)
	}
}

func Benchmark(b *testing.B) {
	for _, ds := range datasets {
		b.Run(ds.name, func(b *testing.B) {
			b.Run("Unmarshal", func(b *testing.B) {
				for i := 0; i < b.N; i++ {
					// ...
				}
			})
			b.Run("Marshal", func(b *testing.B) {
				for i := 0; i < b.N; i++ {
					// ...
				}
			})
			b.Run("Size", func(b *testing.B) {
				for i := 0; i < b.N; i++ {
					// ...
				}
			})
			b.Run("Clone", func(b *testing.B) {
				for i := 0; i < b.N; i++ {
					// ...
				}
			})
			b.Run("Merge", func(b *testing.B) {
				for i := 0; i < b.N; i++ {
					// ...
				}
			})
		})
	}
}
```

Reply:
Thank you very much for this suggestion! This is much cleaner now.

Before I tried to use testing.M.run for several times, but it seems that it shouldn't be, and I didn't realize testing.B.run can run for several times.

I keep most of your code, except I still keep the function for generating the newMessage function as we may add some generator for generating benchmark code in the future so that it can use user defined schema. To write it in an function is easier for generator to change this part of code.

And I still left to deal with one message in one loop currently and I'll discuss this with Feng first.